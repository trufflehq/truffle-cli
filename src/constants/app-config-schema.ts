import Joi from 'joi';

export const CONTENT_PAGE_TYPES = [
  'appManagement',
  'quickActions',
  'youtube',
  'youtubeVideo',
  'youtubeLive',
  'youtubeLiveUpcoming',
  'youtubeLiveNow',
  'youtubeLiveVod',
  'youtubeChatFrame',
  'twitch',
  'twitchEmbed',
] as const;
export type ContentPageType = (typeof CONTENT_PAGE_TYPES)[number];

export const OPERATION_TYPES = [
  'webhook',
  'workflow',
  'exchange',
  'apply-powerup',
  // TODO: support these
  // 'conditional',
] as const;
export type OperationType = (typeof OPERATION_TYPES)[number];

export const ASSET_PARTICIPANT_ENTITY_TYPES = [
  'user',
  'org-member',
  'org',
  'company',
] as const;
export type AssetParticipantEntityType =
  (typeof ASSET_PARTICIPANT_ENTITY_TYPES)[number];

const COUNTABLE_SCHEMA = Joi.object({
  slug: Joi.string().required(),
  name: Joi.string().optional(),
  category: Joi.string().optional(),
  decimalPlaces: Joi.number().optional(),
});

export const EMBED_SCHEMA = Joi.object({
  slug: Joi.string().required(),
  url: Joi.string().required(),
  iconUrl: Joi.string().optional(),
  contentPageType: Joi.string()
    .valid(...CONTENT_PAGE_TYPES)
    .required(),
  windowProps: Joi.object({
    title: Joi.string().optional(),
    tooltipDescription: Joi.string().optional(),
    initialDimensions: Joi.object({
      width: Joi.number().required(),
      height: Joi.number().required(),
    }).optional(),
  }).optional(),
  isResizable: Joi.boolean().optional(),
  resizeBounds: Joi.object({
    minWidth: Joi.number().required(),
    minHeight: Joi.number().required(),
    maxWidth: Joi.number().required(),
    maxHeight: Joi.number().required(),
  }).optional(),
  tileProps: Joi.object({
    bgColor: Joi.string().optional(),
    title: Joi.string().optional(),
  }).optional(),
  defaultStyles: Joi.object().optional(),
  isLoginRequired: Joi.boolean().optional(),
  minTruffleVersion: Joi.string().optional(),
  maxTruffleVersion: Joi.string().optional(),
  deviceType: Joi.string().valid('desktop', 'mobile').optional(),
  parentQuerySelector: Joi.string().optional(),
  status: Joi.string()
    .valid('published', 'experimental', 'disabled')
    .optional(),
});

// TODO: actually implement this in app-config.ts
export const ASSET_SCHEMA = Joi.object({
  path: Joi.string().required(),
  quantity: Joi.number().required(),
});

const POWERUP_SCHEMA = Joi.object({
  slug: Joi.string().required(),
  name: Joi.string().optional(),
  data: Joi.object().optional(),
  imageFileReference: Joi.object().optional(),
});

const ASSET_PARTICIPANT_TEMPLATE_SCHEMA = Joi.object({
  entityType: Joi.string()
    .valid(...ASSET_PARTICIPANT_ENTITY_TYPES)
    .required(),
  entityId: Joi.string().required(),
  share: Joi.number().required(),
});

const ASSET_TEMPLATE_SCHEMA = Joi.object({
  entityType: Joi.string().valid('countable', 'fiat'),
  entityId: Joi.string(),
  entityPath: Joi.string(),
  count: Joi.alternatives()
    .try(
      Joi.number().required(),
      Joi.string().valid('{{USE_PROVIDED}}').required(),
    )
    .required(),
  metadata: Joi.object().optional(),
  senders: Joi.array().items(ASSET_PARTICIPANT_TEMPLATE_SCHEMA).required(),
  receivers: Joi.array().items(ASSET_PARTICIPANT_TEMPLATE_SCHEMA).required(),
})
  // this makes it so that either entityId and entityType is required or entityPath is required
  .with('entityId', 'entityType')
  .xor('entityPath', 'entityId');

export const ACTION_SCHEMA = Joi.object({
  operation: Joi.string()
    .valid(...OPERATION_TYPES)
    .required(),

  // webhook inputs
  url: Joi.when('operation', {
    is: 'webhook',
    then: Joi.string().required(),
  }),

  data: Joi.when('operation', {
    is: 'webhook',
    then: Joi.alternatives()
      .try(Joi.object(), Joi.string().valid('{{USE_PROVIDED}}'))
      .optional(),
  }),

  // workflow inputs
  strategy: Joi.when('operation', {
    is: 'workflow',
    then: Joi.string().valid('sequential', 'parallel').required(),
  }),

  actions: Joi.when('operation', {
    is: 'workflow',
    then: Joi.array().items(Joi.link('#actionSchema')).required(),
  }),

  // exchange inputs
  assets: Joi.when('operation', {
    is: 'exchange',
    then: Joi.alternatives()
      .try(
        Joi.string().valid('{{USE_SECURE_PROVIDED}}').required(),
        Joi.array().items(ASSET_SCHEMA).required(),
        Joi.array().items(ASSET_TEMPLATE_SCHEMA).required(),
      )
      .required(),
  }),

  // apply-pwerup inputs
  powerup: Joi.when('operation', {
    is: 'apply-powerup',
    then: Joi.alternatives().try(Joi.string(), POWERUP_SCHEMA).required(),
  }),

  targetType: Joi.when('operation', {
    is: 'apply-powerup',
    then: Joi.string().required(),
  }),

  targetId: Joi.when('operation', {
    is: 'apply-powerup',
    then: Joi.string().required(),
  }),

  ttlSeconds: Joi.when('operation', {
    is: 'apply-powerup',
    then: Joi.number().required(),
  }),

  inputsTemplate: Joi.object().optional(),

  isDirectExecutionAllowed: Joi.boolean().optional(),
}).id('actionSchema');

const ACTION_WITH_SLUG_SCHEMA = ACTION_SCHEMA.concat(
  Joi.object({
    slug: Joi.string().required(),
  }),
);

export const PRODUCT_VARIANT_SCHEMA = Joi.object({
  slug: Joi.string().required(),
  name: Joi.string().optional(),
  price: Joi.number().required(),
  description: Joi.string().optional(),
  action: ACTION_SCHEMA.optional(),
  assets: Joi.array().items(ASSET_SCHEMA).optional(),
});

export const PRODUCT_SCHEMA = Joi.object({
  slug: Joi.string().required(),
  name: Joi.string().optional(),
  category: Joi.string().optional(),
  rank: Joi.number().optional(),
  variants: Joi.array().items(PRODUCT_VARIANT_SCHEMA).required(),
});

export const APP_CONFIG_SCHEMA = Joi.object({
  path: Joi.string().required(),
  name: Joi.string().required(),
  cliVersion: Joi.string().required(),
  description: Joi.string().optional(),
  embeds: Joi.array().items(EMBED_SCHEMA).optional(),
  countables: Joi.array().items(COUNTABLE_SCHEMA).optional(),
  products: Joi.array().items(PRODUCT_SCHEMA).optional(),
  actions: Joi.array().items(ACTION_WITH_SLUG_SCHEMA).optional(),
  powerups: Joi.array().items(POWERUP_SCHEMA).optional(),
  postInstallAction: Joi.alternatives()
    .try(Joi.string(), ACTION_SCHEMA)
    .optional(),
});
