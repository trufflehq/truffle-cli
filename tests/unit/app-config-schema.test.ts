import { it, describe, expect } from 'vitest';
import * as schemas from '../../src/constants/app-config-schema';
import Joi from 'joi';

describe('app-config-schema', () => {
  describe('EMBED_SCHEMA', () => {
    it('should allow the minimum embed config', () => {
      const embed = {
        slug: 'test-embed',
        url: 'https://example.com',
        contentPageType: 'youtubeVideo',
      };

      Joi.assert(embed, schemas.EMBED_SCHEMA);
    });

    it('should throw an error if the embed config is empty', () => {
      expect(() =>
        Joi.assert({}, schemas.EMBED_SCHEMA),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error if contentPageType is invalid', () => {
      const embed = {
        slug: 'test-embed',
        url: 'https://example.com',
        contentPageType: 'invalidType',
      };

      expect(() =>
        Joi.assert(embed, schemas.EMBED_SCHEMA),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should allow the `windowProps` field', () => {
      const embed = {
        slug: 'test-embed',
        url: 'https://example.com',
        contentPageType: 'youtubeVideo',
        windowProps: {},
      };

      Joi.assert(embed, schemas.EMBED_SCHEMA);
    });

    it('should allow any styles in `defaultStyles`', () => {
      const embed = {
        slug: 'test-embed',
        url: 'https://example.com',
        contentPageType: 'youtubeVideo',
        defaultStyles: {
          backgroundColor: 'red',
          color: 'white',
        },
      };

      Joi.assert(embed, schemas.EMBED_SCHEMA);
    });
  });

  describe('PRODUCT_SCHEMA', () => {
    it('should allow the minimum product config', () => {
      const product = {
        slug: 'test-product',
        variants: [
          {
            slug: 'test-variant',
            price: 10,
          },
        ],
      };

      Joi.assert(product, schemas.PRODUCT_SCHEMA);
    });

    it('should allow a product variant with an action', () => {
      const product = {
        slug: 'test-product',
        variants: [
          {
            slug: 'test-variant',
            price: 10,
            action: {
              operation: 'webhook',
              url: 'https://example.com/webhook',
            },
          },
        ],
      };

      Joi.assert(product, schemas.PRODUCT_SCHEMA);
    });

    it('should throw an error if an action is defined without a url', () => {
      const product = {
        slug: 'test-product',
        variants: [
          {
            slug: 'test-variant',
            price: 10,
            action: {
              operation: 'webhook',
            },
          },
        ],
      };

      expect(() =>
        Joi.assert(product, schemas.PRODUCT_SCHEMA),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should allow a product with `assets` defined', () => {
      const product = {
        slug: 'test-product',
        variants: [
          {
            slug: 'test-variant',
            price: 10,
            assets: [
              {
                path: '@truffle/test-app/_Powerup/test-powerup',
                quantity: 1,
              },
            ],
          },
        ],
      };

      Joi.assert(product, schemas.PRODUCT_SCHEMA);
    });
  });

  describe('ACTION_SCHEMA', () => {
    it('should validate a webhook action with {{USE_PROVIDED}} data', () => {
      const action = {
        operation: 'webhook',
        url: 'https://example.com/webhook',
        data: '{{USE_PROVIDED}}',
      };

      const { error } = schemas.ACTION_SCHEMA.validate(action);
      expect(error).toBeUndefined();
    });

    it('should validate a webhook action with object data', () => {
      const action = {
        operation: 'webhook',
        url: 'https://example.com/webhook',
        data: {
          some: 'data',
        },
      };

      const { error } = schemas.ACTION_SCHEMA.validate(action);
      expect(error).toBeUndefined();
    });

    it('should throw an error if data is an invalid string', () => {
      const action = {
        operation: 'webhook',
        url: 'https://example.com/webhook',
        data: 'invalid',
      };

      const { error } = schemas.ACTION_SCHEMA.validate(action);
      expect(error).toMatchSnapshot();
    });

    it('should validate sub-actions in a workflow', () => {
      const action = {
        operation: 'workflow',
        strategy: 'sequential',
        actions: [
          {
            operation: 'webhook',
            url: 'https://example.com/webhook',
          },
          {
            operation: 'webhook',
            url: 'https://example.com/webhook',
          },
        ],
      };

      const { error } = schemas.ACTION_SCHEMA.validate(action);
      expect(error).toBeUndefined();
    });

    it('should throw an error if a sub-action is invalid', () => {
      const action = {
        operation: 'workflow',
        strategy: 'sequential',
        actions: [
          {
            operation: 'webhook',
          },
          {
            operation: 'webhook',
            url: 'https://example.com/webhook',
          },
        ],
      };

      const { error } = schemas.ACTION_SCHEMA.validate(action);
      expect(error).toMatchSnapshot();
    });

    it('validate an apply-powerup action with a path', () => {
      const action = {
        operation: 'apply-powerup',
        powerup: '@truffle/test-app/_Powerup/test-powerup',
        targetType: 'chat',
        targetId: 'chatId',
        ttlSeconds: 60,
      };

      Joi.assert(action, schemas.ACTION_SCHEMA);
    });

    it('should validate an apply powerup action with a powerup object', () => {
      const action = {
        operation: 'apply-powerup',
        powerup: {
          slug: 'test-powerup',
          name: 'Test Powerup',
        },
        targetType: 'chat',
        targetId: 'chatId',
        ttlSeconds: 60,
      };

      Joi.assert(action, schemas.ACTION_SCHEMA);
    });

    it('should throw an error if an apply-powerup action is invalid', () => {
      const action = {
        operation: 'apply-powerup',
        targetType: 'chat',
        targetId: 'chatId',
        ttlSeconds: 60,
      };

      expect(() =>
        Joi.assert(action, schemas.ACTION_SCHEMA),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('APP_CONFIG_SCHEMA', () => {
    it('should allow a path to be defined for postInstallAction', () => {
      const config = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '1.0.0',
        postInstallAction: '@truffle/app/_Action/test-action',
      };

      Joi.assert(config, schemas.APP_CONFIG_SCHEMA);
    });

    it('should allow an action to be defined for postInstallAction', () => {
      const config = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '1.0.0',
        postInstallAction: {
          operation: 'webhook',
          url: 'https://example.com/webhook',
        },
      };

      Joi.assert(config, schemas.APP_CONFIG_SCHEMA);
    });

    it('should throw an error if postInstallAction is invalid', () => {
      const config = {
        path: '@truffle/test-app',
        name: 'test-app',
        cliVersion: '1.0.0',
        postInstallAction: {
          operation: 'webhook',
        },
      };

      expect(() =>
        Joi.assert(config, schemas.APP_CONFIG_SCHEMA),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
