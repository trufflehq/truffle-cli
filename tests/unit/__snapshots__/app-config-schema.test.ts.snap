// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app-config-schema > ACTION_SCHEMA > should throw an error if a sub-action is invalid 1`] = `[ValidationError: "actions[0].url" is required]`;

exports[`app-config-schema > ACTION_SCHEMA > should throw an error if an apply-powerup action is invalid 1`] = `
[ValidationError: {
  "operation": "apply-powerup",
  "targetType": "chat",
  "targetId": "chatId",
  "ttlSeconds": 60,
  [41m"powerup"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "powerup" is required[0m]
`;

exports[`app-config-schema > ACTION_SCHEMA > should throw an error if data is an invalid string 1`] = `[ValidationError: "data" must be one of [object, {{USE_PROVIDED}}]]`;

exports[`app-config-schema > APP_CONFIG_SCHEMA > should throw an error if postInstallAction is invalid 1`] = `
[ValidationError: {
  "path": "@truffle/test-app",
  "name": "test-app",
  "cliVersion": "1.0.0",
  "postInstallAction": {
    "operation": "webhook",
    [41m"url"[0m[31m [1]: -- missing --[0m
  }
}
[31m
[1] "postInstallAction.url" is required[0m]
`;

exports[`app-config-schema > EMBED_SCHEMA > should throw an error if contentPageType is invalid 1`] = `
[ValidationError: {
  "slug": "test-embed",
  "url": "https://example.com",
  "contentPageType" [31m[1][0m: "invalidType"
}
[31m
[1] "contentPageType" must be one of [appManagement, quickActions, youtube, youtubeVideo, youtubeLive, youtubeLiveUpcoming, youtubeLiveNow, youtubeLiveVod, youtubeChatFrame, twitch, twitchEmbed][0m]
`;

exports[`app-config-schema > EMBED_SCHEMA > should throw an error if the embed config is empty 1`] = `
[ValidationError: {
  [41m"slug"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "slug" is required[0m]
`;

exports[`app-config-schema > PRODUCT_SCHEMA > should throw an error if an action is defined without a url 1`] = `
[ValidationError: {
  "slug": "test-product",
  "variants": [
    {
      "slug": "test-variant",
      "price": 10,
      "action": {
        "operation": "webhook",
        [41m"url"[0m[31m [1]: -- missing --[0m
      }
    }
  ]
}
[31m
[1] "variants[0].action.url" is required[0m]
`;
